server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateNodeRemovalIT.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster;
+
+import org.elasticsearch.action.admin.cluster.health.ClusterHealthResponse;
+import org.elasticsearch.action.admin.cluster.node.shutdown.NodesRemovalPrevalidation;
+import org.elasticsearch.action.admin.cluster.node.shutdown.PrevalidateNodeRemovalAction;
+import org.elasticsearch.action.admin.cluster.node.shutdown.PrevalidateNodeRemovalRequest;
+import org.elasticsearch.action.admin.cluster.node.shutdown.PrevalidateNodeRemovalResponse;
+import org.elasticsearch.cluster.health.ClusterHealthStatus;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.test.ESIntegTestCase;
+
+import static org.hamcrest.Matchers.equalTo;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class PrevalidateNodeRemovalIT extends ESIntegTestCase {
+
+    public void testNodeRemovalFromNonRedCluster() throws Exception {
```

Comment:
While revisiting this test, I thought of two corner cases which are not tested here, and I am not sure they make sense or not. So I'd like to ask for an opinion.

1. If the cluster is green/yellow, and I send a prevalidation request for all nodes in the cluster, the answer would be YES for the overall result and each node's result! Clearly, we do not mean one could remove all those nodes at once w/o causing any issue, even though such a test would pass! Should we explicitly mention this in the doc, or somehow in case of a non-RED cluster just return no responses. From what I understood this API is for clusters which are RED and could be preceded by a call to the cluster health API. I am more inclined towards clarifying what an overall yes means in the doc (specifically pointing out that it doesn't mean that all nodes could be removed at once).
2. Would it be necessary to validate that all the nodes that are specified are data nodes? Currently, we kind of assume that.

Reply:
Thanks for those explicit questions:

1. Yes, we should clarify that in docs. It is not a replacement for using the shutdown API, it is a pre-validation, answering whether doing so would/should succeed. If there are cases here that are not handled by the shutdown API (might be), we should probably also bake them in (as a follow-up).
2. pre-validation should say yes to any non-data node regardless of shard status. We should still encourage to use the pre-validation API in case we have future requirements for other types of nodes.

Reply:
@henningandersen thanks for the answer. I've added some notes to the docs. The PR is ready for review. 

https://elasticsearch_88952.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/prevalidate-node-removal-api.html